src/actions/tamanu/upgrade.rs

```diff
@@ -0,0 +1,192 @@
+	/// By default, this command looks for the most recent installed version of Tamanu.
+	/// If both central and facility servers are present and
+	/// configured, it will pick one arbitrarily.
+	#[arg(short, long)]
+	pub kind: Option<ApiServerKind>,
+
+	/// Path to the Caddyfile.
+	#[arg(long, default_value = DEFAULT_CADDYFILE_PATH)]
+	pub caddyfile_path: PathBuf,
+}
+
+pub async fn run(ctx: Context<TamanuArgs, UpgradeArgs>) -> Result<()> {
+	let UpgradeArgs {
+		kind,
+		version,
+		caddyfile_path,
+	} = ctx.args_sub;
+
+	let (_, root) = find_tamanu(&ctx.args_top)?;
+	let new_root = root.parent().unwrap().join(format!("release-v{version}"));
+
+	let kind = match kind {
+		Some(kind) => kind,
+		None => find_package(&root)?,
+	};
+	info!(?kind, "using");
+
+	if version < Version::parse("2.0.0").unwrap() {
+		bail!("version is too low, bestool doesn't support Tamanu <2.0.0");
+	}
+
+	if !new_root.exists() || !new_root.join(UPGRADED_SIGNAL_NAME).exists() {
+		bail!("new Tamanu not found, try running `prepare-upgrade`");
+	}
+
+	// Caddyfile does not support automation, and this would be easier with JSON.
+	// It's rewriting the config files however since we already use Caddyfile and automated parts are not large.
+	info!(?caddyfile_path, "updating Caddy config file");
+	let re = RegexBuilder::new(r#"c:\\tamanu\\tamanu-web-[0-9]\.[0-9]\.[0-9]"#)
+		.case_insensitive(true)
+		.build()
+		.unwrap();
+	let old_caddyfile = fs::read_to_string(&caddyfile_path).into_diagnostic()?;
+	let new_caddyfile =
+		re.replace_all(&old_caddyfile, format!(r#"C:\tamanu\tamanu-web-{version}"#));
+	fs::write(&caddyfile_path, new_caddyfile.as_bytes()).into_diagnostic()?;
+
+	// Caddy reccomend against using config files and the API at the same time.
+	// This uses the API anyways because this is only a temporary change.
+	// See also https://caddyserver.com/docs/getting-started#api-vs-config-files
+	info!("switch Caddy to upgrading mode");
+
+	let caddyjson = reqwest::Client::new()
+		.post("http://localhost:2019/adapt")
+		.header("Content-Type", "text/caddyfile")
+		.body(new_caddyfile.into_owned())
+		.send()
+		.await
+		.into_diagnostic()?
+		.json::<serde_json::Value>()
+		.await
+		.into_diagnostic()?;
+	let caddyjson = caddyjson
+		.get("result")
+		.ok_or_else(|| miette!("unexpected responce body from Caddy API"))?;
+
+	let mut caddyjson_upgrading = caddyjson.clone();
+	caddyjson_upgrading
+		.pointer_mut("/apps/http/servers/srv0/routes/0/handle/0/routes")
+		.and_then(|v| v.as_array_mut())
+		.ok_or_else(|| miette!("failed to parse Caddy config"))?
+		.insert(
+			1,
+			serde_json::json!({
+				"handle": [
+					{
+						"error": "Upgrading",
+						"handler": "error",
+						"status_code": 526
+					}
+				],
+				"match": [
+					{
+						"not": [
+							{
+								"remote_ip": {
+									"ranges": [
+										"192.168.0.0/16",
+										"172.16.0.0/12",
+										"10.0.0.0/8",
+										"127.0.0.1/8",
+										"fd00::/8",
+										"::1"
+									]
```

Comment:
what's this part for?

Reply:
The general idea is that I want to block public access while allowing private access for health check later. Without the matcher, even `http://localhost` returns 526. This is equivalent of

```caddyfile
@public not remote_ip private_ranges
error @public "Upgrading" 526
```

in Caddyfile. 

(Ok, the above argument is a bit misleading. `http://localhost` returned 526 only when I added the above Caddyfile snippet. Running `upgrade` never blocked `http://localhost`. I'm suspecting this is because the code is only adding the `error` directive to the routes for port 443 while the Caddyfile version adds that to both 80 and 443. I still left this matcher because I didn't want to rely on the difference between http and https. The fact that that works without the matcher is just a byproduct while the matcher is the direct solution.)

Reply:
Yeah, the issue is that Tamanu facility servers are frequently installed "inside the LAN", so that clients will have LAN addresses (like in the 10/8, 192.168/16 CIDRs), so they won't get restricted.

Reply:
Right. There's actually a thing that I don't quite understand where running `curl https://central-demo.palau-dev.tamanu.io` in the terminal of Palau Demo Central returned 526 with the matcher enabled, so I'm not sure if that's the case.